### PR TITLE
fix(connection): delete vpc endpoint when drop connection

### LINKED ISSUE #10450: Connection: DROP Connection should also delete the AWS endpoint

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

close #10450 

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/meta/src/rpc/cloud_provider.rs [hunk 1]
```diff
@@ -46,6 +46,26 @@ impl AwsEc2Client {
...
+    pub async fn delete_vpc_endpoint(&self, vpc_endpoint_id: &str) -> MetaResult<()> {
+        let output = self
+            .client
+            .delete_vpc_endpoints()
+            .vpc_endpoint_ids(vpc_endpoint_id)
+            .send()
+            .await?;
+
+        if let Some(ret) = output.unsuccessful() {
+            if ret.len() > 0 {
```
Comment (ice1000) [inline src/meta/src/rpc/cloud_provider.rs:58]:
```suggestion
            if !ret.is_empty() {
```

**Comment by ice1000** [comment]
CI still fails. Ridiculous.

**Comment by StrikeW** [comment]
> CI still fails. Ridiculous.

Due to openssl vulnerability. :(

**Review by tabVersion — APPROVED** [review]
Basically LGTM, please test carefully before merging

**Comment by ice1000** [comment]
> > CI still fails. Ridiculous.
> 
> Due to openssl vulnerability. :(

#10456 is merged! ;)

**Comment by yezizp2012** [comment]
Should also handle it when drop database. Rest LGTM.

File: src/meta/src/rpc/service/ddl_service.rs [hunk 2]
```diff
@@ -651,12 +651,24 @@ where
...
             .ddl_controller
-            .run_command(DdlCommand::DropConnection(req.connection_id))
+            .run_command(DdlCommand::DropConnection(connection.id))
             .await?;
 
+        // delete AWS vpc endpoint
+        if let Some(connection::Info::PrivateLinkService(svc)) = connection.info
+            && svc.get_provider()? == PbPrivateLinkProvider::Aws {
+            if let Some(aws_cli) = self.aws_client.as_ref() {
+                aws_cli.delete_vpc_endpoint(&svc.endpoint_id).await?;
```
Comment (yezizp2012) [inline src/meta/src/rpc/service/ddl_service.rs:668]: If deleting vpc endpoint fails here, the frontend will receive an error but the connection is already dropped, which will make the ddl operation unretriable. Shall we just print a warning here and leave it to be handled somewhere else maybe in cloud side. 🤔

Comment (yezizp2012) [reply src/meta/src/rpc/service/ddl_service.rs:668]: It also could happen when meta crashes before deleting the endpoint.

Comment (StrikeW) [reply src/meta/src/rpc/service/ddl_service.rs:668]: I also notice that we don't provide atomic semantic to users here. If Meta crashed or the delete vpc operation failed, the vpc endpoint will be leaked and need to be deleted manually. So it is a best-effort semantic. Is it acceptable to you? @mikechesterwang @wjf3121

Comment (yezizp2012) [reply src/meta/src/rpc/service/ddl_service.rs:668]: I found that @mikechesterwang have a suggestion in the slack channel that is to add a `tags` field for connection, which will be used for periodic resource leak check in cloud side. That mechanism LGTM. We can implement it in the next PR.

**Pull request merged**